src/treemodel/schema.js

```diff
@@ -0,0 +1,417 @@
+/**
+ * @license Copyright (c) 2003-2016, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+import clone from '../lib/lodash/clone.js';
+import CKEditorError from '../ckeditorerror.js';
+
+/**
+ * SchemaItem is a singular registry item in {@link core.treeModel.Schema} that groups and holds allow/disallow rules for
+ * one entity. This class is used internally in {@link core.treeModel.Schema} and should not be used outside it.
+ *
+ * @see core.treeModel.Schema
+ * @memberOf core.treeModel
+ */
+export class SchemaItem {
+	/**
+	 * Creates SchemaItem instance.
+	 *
+	 * @param {core.treeModel.Schema} schema Schema instance that owns this item.
+	 */
+	constructor( schema ) {
+		/**
+		 * Schema instance that owns this item.
+		 *
+		 * @member {core.treeModel.Schema} core.treeModel.SchemaItem#_schema
+		 * @private
+		 */
+		this._schema = schema;
+
+		/**
+		 * Paths in which the entity, represented by this item, is allowed.
+		 *
+		 * @member {Array} core.treeModel.SchemaItem#_allowed
+		 * @private
+		 */
+		this._allowed = [];
+
+		/**
+		 * Paths in which the entity, represented by this item, is disallowed.
+		 *
+		 * @member {Array} core.treeModel.SchemaItem#_disallowed
+		 * @private
+		 */
+		this._disallowed = [];
+	}
+
+	/**
+	 * Allows entity, represented by this item, to be in given path.
+	 *
+	 * @param {Array.<String>|String} path Path in which entity is allowed. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 */
+	addAllowed( path, attribute ) {
+		this._addPath( '_allowed', path, attribute );
+	}
+
+	/**
+	 * Disallows entity, represented by this item, to be in given path.
+	 *
+	 * @param {Array.<String>|String} path Path in which entity is disallowed. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 */
+	addDisallowed( path, attribute ) {
+		this._addPath( '_disallowed', path, attribute );
+	}
+
+	/**
+	 * Adds path to the SchemaItem instance.
+	 *
+	 * @param {String} member Name of the array member into which the path will be added. Possible values are `_allowed` or `_disallowed`.
+	 * @param {Array.<String>|String} path Path to be added. String with item names separated by spaces may be passed.
+	 * @param {String} [attribute] If set, this path will be used only for entities that have an attribute with this key.
+	 * @private
+	 */
+	_addPath( member, path, attribute ) {
+		if ( typeof path === 'string' ) {
+			path = path.split( ' ' );
+		}
+
+		path = path.slice();
```

Comment:
Okay. So the rule I go by is that whenever a class method takes an array as a parameter and stores it in class member we should consider cloning the array instead saving the reference to the passed array. Of course there are some cases where you would want to store the reference, but in most cases its best to avoid the situation where class member can be edited without accessing a class at all. It opens up possibilities for mistakes.
```
let myArr = [ 1, 2, 3 ];
let myObj = newMyObj( myArr );
// ... a lot of code
myArr.push( 4 ); // do you really want this to silently impact myObj?
```

Of course when it comes to adding paths to `SchemaItem`s those arrays will mostly be anonymous, but I thought it will be better to be safe than sorry.